File: tests/test_nontof_fwd_back.py
import math
import numpy as np
import parallelproj_backend as pp

from types import ModuleType

from .config import pytestmark


def test_forward_and_back_projection(xp: ModuleType, dev: str):
    """Test the forward projection."""

    ############################################################################
    # setup data

    img_dim = (2, 3, 4)
    voxsize = xp.asarray([4.0, 3.0, 2.0], dtype=xp.float32, device=dev)
    img_origin = (
        -0.5 * xp.asarray(img_dim, dtype=xp.float32, device=dev) + 0.5
    ) * voxsize

    # Read the image from file
    img = xp.reshape(
        xp.asarray(np.loadtxt("img.txt", dtype=np.float32), device=dev), img_dim
    )

    # Read the ray start and end coordinates from file
    vstart = xp.reshape(
        xp.asarray(np.loadtxt("vstart.txt", dtype=np.float32), device=dev), (2, 5, 3)
    )
    vend = xp.reshape(
        xp.asarray(np.loadtxt("vend.txt", dtype=np.float32), device=dev), (2, 5, 3)
    )

    # Calculate the start and end coordinates in world coordinates
    xstart = vstart * voxsize + img_origin
    xend = vend * voxsize + img_origin

    # Read the expected forward values from file
    expected_fwd_vals = xp.reshape(
        xp.asarray(np.loadtxt("expected_fwd_vals.txt", dtype=np.float32), device=dev),
        xstart.shape[:-1],
    )

    ############################################################################

    # Allocate memory for forward projection results
    img_fwd = xp.zeros(xstart.shape[:-1], dtype=xp.float32, device=dev)

    # Perform forward projection
    pp.joseph3d_fwd(xstart, xend, img, img_origin, voxsize, img_fwd)

    # Check if we got the expected results
    eps = 1e-7
    assert (
        float(xp.max(xp.abs(img_fwd - expected_fwd_vals))) < eps
    ), "Forward projection test failed."

    # Allocate memory for back projection results
    bimg = xp.zeros(img_dim, dtype=xp.float32, device=dev)
    ones = xp.ones(xstart.shape[:-1], dtype=xp.float32, device=dev)

    # Perform back projection
    pp.joseph3d_back(xstart, xend, bimg, img_origin, voxsize, ones)

    # Check the results
    ip1 = float(xp.sum(img * bimg))
    ip2 = float(xp.sum(img_fwd * ones))

    eps = 1e-7
    assert abs(ip1 - ip2) / abs(ip1) < eps, "Back projection test failed."


def test_box_projection(xp: ModuleType, dev: str):
    """test forward projection through a uniform box along different axes / angles"""

    # generate an image box full of ones with side length 100mm but non-uniform voxel size
    img_dim = (50, 100, 25)
    voxel_size = xp.asarray([2.0, 1.0, 4.0], device=dev, dtype=xp.float32)
    img_origin = -50 + 0.5 * voxel_size

    # Allocate memory for back projection results
    img = xp.ones(img_dim, dtype=xp.float32, device=dev)

    # create xstart and xend arrays
    xstart = xp.asarray(
        [
            [100, 0, 0],  # exp val 100
            [50, 0, 0],  # exp val 100
            [0, 50, 0],  # exp val 100
            [0, 0, 50],  # exp val 100
            [40, 0, 0],  # exp val 80
            [0, 40, 0],  # exp val 80
            [0, 0, 40],  # exp val 80
            [50, 5, 0],  # exp val sqrt(100^2 + 9^2) = 100.404
            [0, 50, 5],  # exp val sqrt(100^2 + 9^2) = 100.404
            [5, 0, 50],  # exp val sqrt(100^2 + 9^2) = 100.404
            [50, 5, -2],  # exp val sqrt(100^2 + 9^2 + 5^2) = 100.528
            [-2, 50, 5],  # exp val sqrt(100^2 + 9^2 + 5^2) = 100.528
            [5, -2, 50],  # exp val sqrt(100^2 + 9^2 + 5^2) = 100.528
        ],
        device=dev,
        dtype=xp.float32,
    )

    xend = xp.asarray(
        [
            [-100, 0, 0],
            [-50, 0, 0],
            [0, -50, 0],
            [0, 0, -50],
            [-40, 0, 0],
            [0, -40, 0],
            [0, 0, -40],
            [-50, -4, 0],
            [0, -50, -4],
            [-4, 0, -50],
            [-50, -4, 3],
            [3, -50, -4],
            [-4, 3, -50],
        ],
        device=dev,
        dtype=xp.float32,
    )

    exp_vals = xp.asarray(
        [
            100.0,
            100.0,
            100.0,
            100.0,
            80.0,
            80.0,
            80.0,
            math.sqrt(100**2 + 9**2),  # 100.404
            math.sqrt(100**2 + 9**2),  # 100.404
            math.sqrt(100**2 + 9**2),  # 100.404
            math.sqrt(100**2 + 9**2 + 5**2),  # 100.528
            math.sqrt(100**2 + 9**2 + 5**2),  # 100.528
            math.sqrt(100**2 + 9**2 + 5**2),  # 100.528
        ],
        device=dev,
        dtype=xp.float32,
    )

    img_fwd = xp.zeros(xstart.shape[0], dtype=xp.float32, device = dev)

    # Perform back projection
    pp.joseph3d_fwd(xstart, xend, img, img_origin, voxel_size, img_fwd)

    eps = 1e-5
    assert (
        float(xp.max(xp.abs(img_fwd - exp_vals))) < eps
    ), "Forward box projection test failed."

File: tests/test_nontof_fwd_back.cpp
#include "parallelproj.h"
#include "file_utils.h"
#include <iostream>
#include <chrono>
#include <cmath>
#include <vector>
#include <numeric>

int test_box_projection_cpp()
{
  bool all_passed = true;
  // Generate an image box full of ones with side length 100mm but non-uniform voxel size
  std::vector<float> voxel_size = {2.0f, 1.0f, 4.0f};
  std::vector<int> img_dim = {50, 100, 25};
  std::vector<float> img_origin = {-50.0f + 0.5f * voxel_size[0],
                                   -50.0f + 0.5f * voxel_size[1],
                                   -50.0f + 0.5f * voxel_size[2]};

  // Allocate memory for image
  size_t nvox = img_dim[0] * img_dim[1] * img_dim[2];
  std::vector<float> img(nvox, 1.0f);

  // xstart and xend arrays
  std::vector<std::vector<float> > xstart = {
      {100, 0, 0}, {50, 0, 0}, {0, 50, 0}, {0, 0, 50}, {40, 0, 0}, {0, 40, 0}, {0, 0, 40}, {50, 5, 0}, {0, 50, 5}, {5, 0, 50}, {50, 5, -2}, {-2, 50, 5}, {5, -2, 50}};
  std::vector<std::vector<float> > xend = {
      {-100, 0, 0}, {-50, 0, 0}, {0, -50, 0}, {0, 0, -50}, {-40, 0, 0}, {0, -40, 0}, {0, 0, -40}, {-50, -4, 0}, {0, -50, -4}, {-4, 0, -50}, {-50, -4, 3}, {3, -50, -4}, {-4, 3, -50}};

  std::vector<float> exp_vals = {
      100.0f, 100.0f, 100.0f, 100.0f,
      80.0f, 80.0f, 80.0f,
      static_cast<float>(std::sqrt(100 * 100 + 9 * 9)),
      static_cast<float>(std::sqrt(100 * 100 + 9 * 9)),
      static_cast<float>(std::sqrt(100 * 100 + 9 * 9)),
      static_cast<float>(std::sqrt(100 * 100 + 9 * 9 + 5 * 5)),
      static_cast<float>(std::sqrt(100 * 100 + 9 * 9 + 5 * 5)),
      static_cast<float>(std::sqrt(100 * 100 + 9 * 9 + 5 * 5))};

  size_t n_lors = xstart.size();
  std::vector<float> xstart_flat, xend_flat;
  for (size_t i = 0; i < n_lors; ++i)
  {
    xstart_flat.insert(xstart_flat.end(), xstart[i].begin(), xstart[i].end());
    xend_flat.insert(xend_flat.end(), xend[i].begin(), xend[i].end());
  }

  std::vector<float> img_fwd(n_lors, 0.0f);

  // Call the C++ joseph3d_fwd function
  joseph3d_fwd(
      xstart_flat.data(),
      xend_flat.data(),
      img.data(),
      img_origin.data(),
      voxel_size.data(),
      img_fwd.data(),
      n_lors,
      img_dim.data(),
      0, // device_id
      64 // threadsperblock
  );

  // Check results
  float eps = 1e-5f;
  for (size_t i = 0; i < n_lors; ++i)
  {
    if (std::abs(img_fwd[i] - exp_vals[i]) >= eps)
    {
      std::cerr << "Forward box projection test failed at i=" << i
                << ": got " << img_fwd[i] << ", expected " << exp_vals[i] << std::endl;
      all_passed = false;
    }
  }

  return all_passed ? 0 : 1;
}

int test_forward_and_back_projection_cpp()
{
  ////////////////////////////////////////////////////////
  // host array test cases
  ////////////////////////////////////////////////////////

  bool all_tests_passed = true;

  std::cout << "Host array test\n";

  std::vector<int> img_dim = {2, 3, 4};
  std::vector<float> voxsize = {4.0f, 3.0f, 2.0f};

  std::vector<float> img_origin(3);
  for (int i = 0; i < 3; ++i)
  {
    img_origin[i] = (-(float)img_dim[i] / 2 + 0.5f) * voxsize[i];
  }

  // Read the image from file
  std::vector<float> img = readArrayFromFile<float>("img.txt");

  // Read the ray start coordinates from file
  std::vector<float> vstart = readArrayFromFile<float>("vstart.txt");

  // Read the ray end coordinates from file
  std::vector<float> vend = readArrayFromFile<float>("vend.txt");

  size_t nlors = vstart.size() / 3;

  // Calculate the start and end coordinates in world coordinates
  std::vector<float> xstart(3 * nlors);
  std::vector<float> xend(3 * nlors);

  for (int ir = 0; ir < nlors; ir++)
  {
    xstart[ir * 3 + 0] = img_origin[0] + vstart[ir * 3 + 0] * voxsize[0];
    xstart[ir * 3 + 1] = img_origin[1] + vstart[ir * 3 + 1] * voxsize[1];
    xstart[ir * 3 + 2] = img_origin[2] + vstart[ir * 3 + 2] * voxsize[2];

    xend[ir * 3 + 0] = img_origin[0] + vend[ir * 3 + 0] * voxsize[0];
    xend[ir * 3 + 1] = img_origin[1] + vend[ir * 3 + 1] * voxsize[1];
    xend[ir * 3 + 2] = img_origin[2] + vend[ir * 3 + 2] * voxsize[2];
  }

  // Allocate memory for forward projection results
  std::vector<float> img_fwd(nlors);

  size_t nvoxels = img_dim[0] * img_dim[1] * img_dim[2];

  // Perform forward projection
  joseph3d_fwd(
      xstart.data(), xend.data(), img.data(),
      img_origin.data(), voxsize.data(), img_fwd.data(),
      nlors, img_dim.data(), 0, 64);

  /////////////////////////////////////////////////////////////////////////////
  /////////////////////////////////////////////////////////////////////////////
  /////////////////////////////////////////////////////////////////////////////

  // Read the expected forward values from file
  std::vector<float> expected_fwd_vals = readArrayFromFile<float>("expected_fwd_vals.txt");

  // Check if we got the expected results
  float fwd_diff = 0;
  float eps = 1e-7;

  printf("\nforward projection test\n");
  for (int ir = 0; ir < nlors; ir++)
  {
    printf("test ray %d: fwd projected: %.7e expected: %.7e\n", ir, img_fwd[ir], expected_fwd_vals[ir]);

    fwd_diff = std::abs(img_fwd[ir] - expected_fwd_vals[ir]);
    if (fwd_diff > eps)
    {
      std::cerr << "Forward projection test failed.\n";
      std::cerr << "Difference: " << fwd_diff << "\n";
      std::cerr << "Expected: " << expected_fwd_vals[ir] << "\n";
      std::cerr << "Actual: " << img_fwd[ir] << "\n";
      std::cerr << "Tolerance: " << eps << "\n";
      std::cerr << "Ray index: " << ir << "\n";

      all_tests_passed = false;
    }
  }

  ////////////////////////////////////////////////////////////////////////////
  ////////////////////////////////////////////////////////////////////////////
  ////////////////////////////////////////////////////////////////////////////

  // Test the back projection
  std::vector<float> bimg(img_dim[0] * img_dim[1] * img_dim[2], 0.0f);
  std::vector<float> ones(nlors, 1.0f);

  joseph3d_back(
      xstart.data(), xend.data(), bimg.data(),
      img_origin.data(), voxsize.data(), ones.data(),
      nlors, img_dim.data());

  printf("\nback projection of ones along all rays:\n");
  for (size_t i0 = 0; i0 < img_dim[0]; i0++)
  {
    for (size_t i1 = 0; i1 < img_dim[1]; i1++)
    {
      for (size_t i2 = 0; i2 < img_dim[2]; i2++)
      {
        printf("%.1f ", bimg[img_dim[1] * img_dim[2] * i0 + img_dim[2] * i1 + i2]);
      }
      printf("\n");
    }
    printf("\n");
  }

  // To test whether the back projection is correct, we test if the back projector is the adjoint
  // of the forward projector. This is more practical than checking a lot of single voxels in the
  // back projected image.

  float inner_product1 = std::inner_product(img.begin(), img.end(), bimg.begin(), 0.0f);
  float inner_product2 = std::inner_product(img_fwd.begin(), img_fwd.end(), ones.begin(), 0.0f);

  float ip_diff = fabs(inner_product1 - inner_product2);

  if (ip_diff > eps)
  {
    std::cerr << "Back projection test failed.\n";
    std::cerr << "Inner product 1: " << inner_product1 << "\n";
    std::cerr << "Inner product 2: " << inner_product2 << "\n";
    std::cerr << "Difference: " << ip_diff << "\n";
    std::cerr << "Tolerance: " << eps << "\n";

    all_tests_passed = false;
  }

  return all_tests_passed ? 0 : 1;
}

int main()
{
  int result = 0;
  result |= test_box_projection_cpp();
  result |= test_forward_and_back_projection_cpp();
  return result;
}
